docs/modules/record-encryption/proc-configuring-record-encryption-filter.adoc

```diff
@@ -35,14 +35,21 @@ filters:
       selector: <KEK_selector_service_name> # <3>
       selectorConfig:
         template: "KEK_${topicName}" # <4>
+      experimental:
+        "encryptionDekRefreshAfterWriteSeconds": 3600 # <6>
+        "encryptionDekExpireAfterWriteSeconds": 7200 # <7>
+        "maxEncryptionsPerDek": 5000000 # <8>
 	 # ...
 ----
 <1> The KMS service name.
 <2> Configuration specific to the KMS provider.
 <3> The Key Encryption Key (KEK) selector to use. The `${topicName}` is a literal understood by the proxy. 
 For example, if using the `TemplateKekSelector` with the template `KEK_$\{topicName}`, create a key for every topic that
 is to be encrypted with the key name matching the topic name, prefixed by the string `KEK_`.
-<4> The template for deriving the KEK, based on a specific topic name. 
+<4> The template for deriving the KEK, based on a specific topic name.
+<5> How long to use a DEK for encryption after it has been created, after this time the Filter will asynchronously obtain a new DEK while continuing using the old one. This can be used to control how quickly we respond to the fact that a KEK has been rotated externally. Note that tuning this value has a cost implication. Shorter durations will result in increased operations on the KMS at encrypt and decrypt time, because more DEKs will be created and used.
+<6> How long after creation a DEK is eligible to be evicted from the cache. Only actively used DEKs will be refreshed, this setting puts a bound on how long unused DEKs remain cached.
+<7> The maximum number of records any DEK should be used to encrypt. After this limit is hit, that DEK will be destroyed and a new one created. As above, reducing the value has a cost because we may create more DEKs.
 
```

Comment:
I'd bring the note from the bullet above out into the main body of the doc.

```suggestion
NOTE:  The `encryptionDekRefreshAfterWriteSeconds`, `encryptionDekExpireAfterWriteSeconds` and `maxEncryptionsPerDek` settings govern the lifetime of the DEK.  If the DEK has a shorter lifespan, Kroxylicious must call the KMS to generatenew DEKs or decrypt them more often.  The additional KMS API calls will increase produce/consume latency and may increase your KMS provider costs.   Conversely, if DEKs have a longer lifespan the key material may become exhausted/.
```

Reply:
Have done so, just one question on `Conversely, if DEKs have a longer lifespan the key material may become exhausted`. Does that need elaboration? Are you talking about the user setting the values so high that they exceed the recommended bytes/operations?

Reply:
I think its the same concept of exhaustion, `maxEncryptionsPerDek` and `encryptionDekExpireAfterWriteSeconds` are both meant to ensure we don't push too many bytes through a given DEK. 

I haven't looked in detail about how hard it is do arrange but feels a bit odd that we don't actually track the number bytes pushed through a DEK and provide a limit based on that.

Reply:
Sorry its `encryptionDekRefreshAfterWriteSeconds` that maps to the exhaustion problem, `encryptionDekExpireAfterWriteSeconds` is more about limiting the size of a cache and speeds up only decryptions in the 2nd hour.

Reply:
The difference is "if DEKs have a longer lifespan the key material may become exhausted.", in our code if we decrease the lifespan by setting `maxEncryptionsPerDek` to 1, then the DEK is exhausted very quickly (using the language in our code). But we wouldn't say that we've made it mode likely to exhaust the key material.

Reply:
Yes, sorry, my wording is sloppy, I've conflated the concepts.   Let's fix that.

`maxEncryptionsPerDek` is the control we currently have to prevent key exhaustion.   As Sam says, we ought to add a byte limit (is there an issue?).

`encryptionDekRefreshAfterWriteSeconds` and `encryptionDekExpireAfterWriteSeconds` are controls that limit the 'originator usage period' in NIST terminology.

https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r5.pdf (page 38).

These controls are about limiting the blast radius if a DEK is leaked

So maybe we reword to split into two paragraphs.

```suggestion
`encryptionDekRefreshAfterWriteSeconds` and`encryptionDekExpireAfterWriteSeconds` help govern the "originator usage period" of the DEK. That is the period of time the DEK will be used to encrypt records.  Keeping the period short reduces helps the blast radius in the event that DEK key material is leaked.  However, there is a trade-off.  The additional KMS API calls will increase produce/consume latency and may increase your KMS provider costs.

`maxEncryptionsPerDek` help prevent key exhaustion by placing an upper limit of the amount of times that a DEK may be used to encrypt records.
```

Reply:
> Sorry its encryptionDekRefreshAfterWriteSeconds that maps to the exhaustion problem, encryptionDekExpireAfterWriteSeconds is more about limiting the size of a cache and speeds up only decryptions in the 2nd hour.

Not that it matters much, but I don't see it quite like that.  You could use either to govern the "originator usage period" of the DEK.

The thing that `encryptionDekRefreshAfterWriteSeconds` gives you better producer tail latency control as the DEK refresh occurs in the background.  If you use `encryptionDekExpireAfterWriteSeconds` OTOH, the producer needs to await whilst the KMS generates you a new DEK.

Reply:
issue created for adding a bytes-encrypted limit to DEKs https://github.com/kroxylicious/kroxylicious/issues/1423

Reply:
The NIST docs also have some interesting things to say about usage periods for symmetric wrapping keys
>Cryptoperiod: The recommended originator-usage period for a symmetric key-wrapping
key that is used to wrap very large numbers of keys over a short period of time is on the
order of a day or a week. If a relatively small number of keys are to be wrapped under a
key-wrapping key, the originator-usage period of the key-wrapping key could be up to
two years. In the case of a key-wrapping key that is used for only a single message or
communications session, the cryptoperiod would be limited to a single communication
session. It is recommended that a recipient-usage period extend no more than three years
beyond the end of the originator-usage period.

A fleet of proxies generating DEKs will churn through keys. So while KEK rotation is not our direct concern we should be pointing to it. We should also probably add a metric around DEK generation (per KEK??)